### ISSUE TITLE
dbt - L009: Files must end with a single trailing newline.

### ISSUE THREAD

**Issue by tekumara** [sqlfluff/sqlfluff#2046]
## Expected Behaviour
No error

## Observed Behaviour
```
L:  5 | P:  60 | L009 | Files must end with a single trailing newline.    
```

## Steps to Reproduce
A dbt model that calls a macro, eg:
```
{{ config ( materialized='incremental',
   transient=false,
    ) }}

{{ generate_table('mytable') }}

```

## Dialect
snowflake

## Version
sqlfluff, version 0.8.2
python 3.9

## Configuration
```
[tool.sqlfluff.core]
templater = "dbt"
sql_file_exts = ".sql,.sql.j2,.dml,.ddl"
dialect = "snowflake"
ignore_templated_areas = true

[tool.sqlfluff.indentation]
indented_joins = false
indented_using_on = true
template_blocks_indent = false

[tool.sqlfluff.templater.dbt]
project_dir = "."
profile = "snowflake"
```

**Comment by tekumara** [comment]
Yep that's all.

Here's the tail of `sqlfluff parse`:
```
[L:  8, P:  1]      |                        identifier:                           'RunID'
[L:  8, P: 44]      |                    [META] dedent:
[L:  8, P:  1]      |    newline:                                                  '\n'
[L:  8, P:  1]      |    newline:                                                  '\n'
[L:  8, P: 44]      |    newline:                                                  '\n'
```

Here's the tail of the macro:
```
ORDER BY rl.RunID

{% endmacro %}

```

If I remove the newline before `{% endmacro %}` it produces two newlines in the parse, and still errors.

**Comment by jpy-git** [comment]
Makes sense thanks! This is a known issue with the dbt templater.

In Jinja, the template lines are included as placeholders in the parse tree so it's possible to account for those
e.g.
![image](https://user-images.githubusercontent.com/80432516/145007044-7760e2cc-458f-46de-8468-fe1eac22a9f0.png)

However, as you can see from your parse tree, certain dbt templates don't show up in the parse tree. So as far as the rule can see the file finishes with 3 newlines. We encountered a similar error in L050 (no newlines at start of the file) https://github.com/sqlfluff/sqlfluff/pull/1840#issuecomment-962682902. I ended up adding some logic to infer the presence of those missing template blocks so I'll try and add some similar logic to L009.

I'll also create a separate issue to see if one of the dbt contributors can get those templates to show up in the parse tree as a more long term solution.

**Comment by barrywhart** [comment]
See this [comment](https://github.com/sqlfluff/sqlfluff/issues/2066#issuecomment-988800699) on #2066 for more details on what's happening. TL;DR: This is actually not a dbt templater issue; addressing issues like this systematically will require a parser change.

**Comment by barrywhart** [comment]
Another update: we decided not to change the parser, but the work in PR #2126, once it is merged, will provide nicer ways of working with collections of segments and raw/templated slices, and that's what is needed to fix this issue. (To be clear, it's possible to fix it now, but it'll be much easier then.)

**Comment by barrywhart** [comment]
This issue can be worked on now. I suggest trying to use the new classes and functions in `sqlfluff.core.rules.functional`, which should provide more convenient access & conversions from the parser segments to the related raw and/or templated file slices.

**Comment by barrywhart** [comment]
Reopening because my initial PR didn't handle the case of adding a missing newline to the end of the file if the last code in the file is templated code -- this is turning up some bugs in the core linter that need to be addressed.
